core/src/main/java/org/opensearch/sql/executor/QueryService.java

```diff
@@ -112,11 +112,14 @@ public void executeWithCalcite(
         log.warn("Fallback to V2 query engine since got exception", t);
         executeWithLegacy(plan, queryType, listener, Optional.of(t));
       } else {
-        if (t instanceof Error) {
+        if (t instanceof Exception) {
+          listener.onFailure((Exception) t);
+        } else if (t instanceof VirtualMachineError) {
+          // throw and fast fail the VM errors such as OOM (same with v2).
+          throw t;
+        } else {
```

Comment:
To align with behaviour of v2, OOM error will fail the whole JVM. We caught OOM error in previous implementation.

Reply:
> But will it be better to trigger GC and then check memory usage again before making the service crash?

Maybe more discussion is required and track in a follow-up issue.